internal/metrics/daemon.go

```diff
@@ -0,0 +1,201 @@
+		tg.cancel()
+	}
+
+	tg.createScrapper()
+	tg.ctx, tg.cancel = context.WithCancel(context.Background())
+	ticker := time.NewTicker(tg.interval)
+	tg.lock.Unlock()
+
+	getdata := func() {
+		log.Debugf("ticker run for workload '%v'", tg.workloadName)
+		ctx, _ := context.WithTimeout(context.Background(), scrapeTimeout)
+		data := tg.Run(ctx)
+		cb(data, map[string]string{workloadMetric: tg.workloadName})
+	}
+	for {
+		select {
+		case <-tg.forcetrigger:
+			getdata()
+		case <-ticker.C:
+			getdata()
+		case <-tg.ctx.Done():
+			log.Debugf("time ticker for workload '%v' stopped", tg.workloadName)
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+func (tg *TargetMetric) IsStopped() bool {
+	return tg.cancel == nil
+}
+
+func (tg *TargetMetric) Stop() {
+	tg.lock.Lock()
+	defer tg.lock.Unlock()
+
+	tg.cancel()
+	tg.cancel = nil
+	return
+}
+
+func (tg *TargetMetric) createScrapper() {
+	scrappers := []*HTTPScraper{}
+	for _, url := range tg.urls {
+		scrappe, err := NewHTTPScraper(url)
+		if err != nil {
+			log.Errorf("cannot start scrapper for %v: %v", url, err)
+		}
+		scrappers = append(scrappers, scrappe)
+	}
+	tg.scrappers = scrappers
+}
+
+func (tg *TargetMetric) Run(ctx context.Context) []model.Vector {
+	res := []model.Vector{}
+	for _, scrape := range tg.scrappers {
+		data, err := scrape.Scrape(ctx)
+		if err != nil {
+			log.Errorf("cannot get metrics for workload '%v': %v", tg.workloadName, err)
+			continue
+		}
+		res = append(res, data)
+	}
+	return res
+}
+
+//go:generate mockgen -package=metrics -destination=mock_daemon.go . MetricsDaemon
+type MetricsDaemon interface {
+	Start()
+	AddTarget(workloadName string, urls []string, interval time.Duration, labelsMap map[string]string)
+	DeleteTarget(workloadName string)
+}
+
+type metricsDaemon struct {
+	targets map[string]*TargetMetric
+	lock    sync.RWMutex
+	store   API
+}
+
+func NewMetricsDaemon(store API) *metricsDaemon {
+	return &metricsDaemon{
+		targets: map[string]*TargetMetric{},
+		lock:    sync.RWMutex{},
+		store:   store,
+	}
+}
+
+// GetTargets returns a list of current workload targets
+func (md *metricsDaemon) GetTargets() []string {
+	md.lock.Lock()
+	defer md.lock.Unlock()
+	res := []string{}
+	for key := range md.targets {
+		res = append(res, key)
+	}
+	return res
+}
+
+// Start started the target daemon.
+func (md *metricsDaemon) Start() {
+	md.lock.Lock()
+	defer md.lock.Unlock()
+	for _, target := range md.targets {
+		md.startTarget(target)
+	}
+	return
+}
+
+func (md *metricsDaemon) startTarget(target *TargetMetric) {
+	cb := func(data []model.Vector, labels map[string]string) {
+		for _, vector := range data {
+			err := md.store.AddVector(vector, labels)
+			if err != nil {
+				log.Errorf("cannot store vector: %v", err)
+			}
+		}
+	}
+	go target.Start(cb)
+}
+
+// AddTarget adds a target to  listen for the  given workloadName.
+func (md *metricsDaemon) AddTarget(workloadName string, urls []string, interval time.Duration, labelsMap map[string]string) {
```

Comment:
WDYT about changing `workloadName` to something more generic, like `targetName`? I would like to add system metrics target, which wouldn't have a "workload name".